// sematable/lib/PageSizeContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _selectors = require('./selectors');

var _selectors2 = _interopRequireDefault(_selectors);

var _actions = require('./actions');

var _PageSize = require('./PageSize');

var _PageSize2 = _interopRequireDefault(_PageSize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var mapStateToProps = function mapStateToProps(state, props) {
  var tableName = props.tableName;

  var selectors = (0, _selectors2.default)(tableName);
  var isInitialized = selectors.getIsInitialized(state);

  if (!isInitialized) {
    return {
      isInitialized: isInitialized
    };
  }

  var pageInfo = selectors.getPageInfo(state, props);

  return {
    isInitialized: isInitialized,
    pageSize: pageInfo.pageSize,
    pageSizes: pageInfo.pageSizes,
    value: selectors.getFilter(state),
    options: selectors.getFilterOptions(state),
    totalSize: selectors.getFiltered(state).length
  };
};

var mapDispatchToProps = function mapDispatchToProps(dispatch, _ref) {
  var tableName = _ref.tableName;
  return {
    onChange: function onChange(pageSize) {
      return dispatch((0, _actions.tablePageSizeChanged)(tableName, pageSize));
    }
  };
};

exports.default = (0, _reactRedux.connect)(mapStateToProps, mapDispatchToProps)(function (_ref2) {
  var isInitialized = _ref2.isInitialized,
      otherProps = _objectWithoutProperties(_ref2, ['isInitialized']);

  if (!isInitialized) {
    return null;
  }
  return _react2.default.createElement(_PageSize2.default, otherProps);
});
